lang/py/src/avro/io.py

```diff
@@ -580,16 +696,36 @@ def read_data(self, writers_schema, readers_schema, decoder):
     elif writers_schema.type == 'string':
       return decoder.read_utf8()
     elif writers_schema.type == 'int':
-      return decoder.read_int()
+      if hasattr(writers_schema, 'logical_type'):
+        if writers_schema.logical_type == constants.DATE:
+          return decoder.read_date_from_int()
+        elif writers_schema.logical_type == constants.TIME_MILLIS:
+          return decoder.read_time_millis_from_int()
```

Comment:
```
In [12]: datetime.date(1969, 1, 1) - datetime.date(1970, 1, 1)
Out[12]: datetime.timedelta(-365)

In [13]: delta = datetime.date(1969, 1, 1) - datetime.date(1970, 1, 1)

In [14]: delta.days
Out[14]: -365
```
This date is before epoch time but its still valid and will pass through your `write_date_int(self, datum)` method. And since `-365` is a perfectly valid `int` it will encode it. Its very easy for someone to make a mistake (a typo?) in their JSON. Its normally a good practice to protect the API from being misused is anyway. A simple example would be the `datetime` library itself. We know that months cannot have a value more than 12 but they still protect the API around it.
```
In [15]: datetime.date(1970, 50, 50)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)

/Users/pxu/MyCode/replication_handler/<ipython console> in <module>()

ValueError: month must be in 1..12
```